File: R/drjacoby_LL_timeseries.cpp
cpp_loglike <- "SEXP loglike(Rcpp::NumericVector params, int param_i, Rcpp::List data, Rcpp::List misc) {

  // fixed parameters
  // from Verity et al. 2020 Lancet ID
  double r = 0.14;
  double m_od = 18.8;
  double s_od = 0.45;

  // unpack data
  std::vector<double> dat = Rcpp::as< std::vector<double> >(data[\"obs_death\"]);

  // extract parameters
  double ma = params[\"ma\"];
  double I0 = params[\"I0\"];

  // extract misc objects
  std::vector<double> pa = Rcpp::as< std::vector<double> >(misc[\"pa\"]);
  int min_day = Rcpp::as<int>(misc[\"min_day\"]);
  int curr_day = Rcpp::as<int>(misc[\"curr_day\"]);

  // counter iterms
  int agelen = pa.size();

  // how do I integrate
  std::vector<double> expc;

  // get log-likelihood over all days
  double ret = 0.0;
  for(int d = min_day; d < curr_day; d++){
    for(int a = 0; a < agelen; a++){
    ret += R::dpois(expc[d][a], dat[d][a], true);
    }
  }

  // account for reparameterization
  // ret = ret + (agelen - 1) * log(ma)

  // catch underflow
  if (!std::isfinite(ret)) {
    const double OVERFLO_DOUBLE = DBL_MAX/100.0;
    ret = -OVERFLO_DOUBLE;
  }

  // return as SEXP
  return Rcpp::wrap(ret);
}"



# define cpp logprior function
cpp_logprior <- "SEXP logprior(Rcpp::NumericVector params, int param_i, Rcpp::List misc) {

  // extract parameters
  //double ma = params[\"ma\"];
  double I0 = params[\"I0\"];

  // fixed/derived parameters
  double sigma_mean = 0.7;
  double sigma_var = 1;
  double lnorm_sigma2 = log(sigma_var/(sigma_mean*sigma_mean) + 1.0);
  double lnorm_mu = log(sigma_mean) - lnorm_sigma2/2.0;

  // calculate logprior
  // ma prior is dunif 1 to 10 or -log(9)
  double ret = -log(9) + R::dlnorm(I0, lnorm_mu, sqrt(lnorm_sigma2), true);

  // catch underflow
  if (!std::isfinite(ret)) {
    const double OVERFLO_DOUBLE = DBL_MAX/100.0;
    ret = -OVERFLO_DOUBLE;
  }

  // return as SEXP
  return Rcpp::wrap(ret);
}"

File: R/drjacoby_LL_timeseries.R
#' @title Likelihood for Curve Aware
r_tod_log_like <- function(params, param_i, data, misc) {

  # assume r is fixed
  r <- 0.14
  # alpha and beta for time from infection to time to death
  # from lancet id paper
  m_od <- 18.8
  s_od <- 0.45
  I0 <- 2
  # free params
  ma <- params[1]
  #r1 <- params[2]
  #I0 <- params[2]
  #ma9 <- params[1]
  #r1 <- params[2]
  # r2 <- params[3]
  # r3 <- params[4]
  # r4 <- params[5]
  # r5 <- params[6]
  # r6 <- params[7]
  # r7 <- params[8]
  # r8 <- params[9]
  # I0 <- params[10]
  # scalars <- c(r1, r2, r3, r4, r5, r6, r7, r8)
  #ma <- sapply(scalars, function(x){x * ma9})

  get_exp_deaths_by_day <- function(bigT){

    # expected deaths
    integrand <- function(t, gr = r){ return(
      I0 * exp(gr*t) * pgamma(bigT - t, shape = 1/s_od^2, scale = m_od*s_od^2)) }
    integral <- integrate(integrand, lower = -Inf, bigT)

    # total exp deaths
    exp.deaths <- misc$pa * ma * integral$value
    # out
    return(exp.deaths)
  }

  exp_deaths.day <- lapply(misc$min_day:misc$curr_day, get_exp_deaths_by_day)


  #..................
  # poisson
  #..................
  get_pois <- function(obs_death, exp_death){
    ret <- sum(dpois(x = obs_death, lambda = exp_death, log = T))
    #ret <- ret + length(scalars) * log(ma9) # account for reparameterization
    #ret <- ret + log(ma2) # account for reparameterization
    return(ret)
  }

  loglik <- mapply(get_pois, obs_death = data$obs_deaths, exp_death = exp_deaths.day)
  loglik <- sum(loglik)
  return(loglik)

}

#' @title Prior for Curve Aware
r_tod_log_prior <- function(params, param_i, misc) {

  ma <- params[1]
  #r1 <- params[2]
  #I0 <- params[2]
  #ma9 <- params[1]
  #r1 <- params[2]
  # r2 <- params[3]
  # r3 <- params[4]
  # r4 <- params[5]
  # r5 <- params[6]
  # r6 <- params[7]
  # r7 <- params[8]
  # r8 <- params[9]
  # I0 <- params[10]
  # scalars <- c(r1, r2, r3, r4, r5, r6, r7, r8)
  # flat prior
  dayprior <- sum(
                  sapply(misc$min_day:misc$curr_day, function(dummy){
                   # ret <- dunif(ma9, min = 0, max = 1, log = TRUE) +
                   #   sum( sapply(scalars, function(x){dlnorm(x, meanlog = 0, sdlog = 5, log = TRUE)}) )
                    #+ dunif(I0, min = 0, max = 10, log = TRUE)
                    ret <- dunif(ma, min = 0, max = 1, log = TRUE)
                      #dlnorm(r1, meanlog = 0, sdlog = 5, log = TRUE) +
                      #+ dlnorm(I0, meanlog = 0.7, sd = 0.05, log = TRUE)

                    return(ret)
                  })
              )
  return(dayprior)
}



